sequtils: simplify `mapIt` implementation

zerbina opened the pull request:
## Summary

Use an inline iterator instead of a macro for achieving the "evaluate
once" semantics in the `mapIt` template. This makes the code simpler,
and, given that the compiler's code itself uses `mapIt`, also reduces
the amount of macro usage during bootstrapping.

## Details

* introduce the internal `mapIter` iterator
* remove the now obsolete internal `evalOnceAs` macro
* adjust the internal documentation of `mapIt` a bit
* adjust the `t8339.nim` test to not depend on `sequtils`
* don't introduce an unnecessary shallow copy when lifting `op` into a
  procedure

While the `mapIter` doesn't use `lent`, special care is taken for it to
be eligible for the cursor optimization. With the current compiler, for-
vars are always turned into cursors when the invoked inline iterator
only ever yields the *symbol* of a local.

The `t8339.nim` test is intended to make sure that aliases are skipped
when rendering types, but using `sequtils` for testing that is not
required, and only made the test unnecessarily depend on the `mapIt`
implementation details (which now changed, making the test fail).

saem commented on lib/pure/collections/sequtils.nim:
```diff
@@ -991,46 +980,34 @@ template mapIt*(s: typed, op: untyped): untyped =
+      block:
+        var it{.inject.}: InType;
+        op), typeOfProc)
+  # we cannot use normal overloading instead of the 'when' here, as the that
+  # force `op` to be ``typed``, which is what we don't want it to be
```
```suggestion
  # we use `when` instead of overload resolution which would require `op` to be
  # ``typed``, allowing optimization of some `op`s to inline expressions
  # instead of requiring lambdas.
```

Bit more verbose, but I believe it accurately captures the intention (don't assume the `when` is bad) and convey the impact of mucking about with it carelessly.

zerbina replied:
Hm, I was trying to communicate that it'd be better for there to be two overloads:
```nim
template mapIt(s: typed, op: proc) = ...
template mapIt(s: typed, op: untyped) = ...
```
instead of implementing manual overloading via `when` (which is only a workaround).

saem replied:
Good thing I wrote out my interpretation.

Could also do:
```nim
  # xxx: drop the `when` by overloading `mapIt`, one where the `op` arg is of
  #      type `proc` and the other `untyped`
```

Again, all soft suggestions, not blockers.

zerbina replied:
Yep, that is better, thanks. I've turned the comment into an `xxx` one, with an additional note that a compiler issue needs to be resolved first.

saem commented on lib/pure/collections/sequtils.nim:
```diff
@@ -991,46 +980,34 @@ template mapIt*(s: typed, op: untyped): untyped =
-        # (`error: use of undeclared identifier`) instead of Nim compile errors
-        evalOnceAs(s2, s, compiles((let _ = s)))
-
+    # the normal version
```
```suggestion
    # the normal version (inline expression)
```

saem reviewed (approved):
soft suggestions.

That's a fair bit easier to read! 👓

zerbina committed:
sequtils: simplify `mapIt` implementation

Summary
=======

Use an inline iterator instead of a macro for achieving the "evaluate
once" semantics in the `mapIt` template. This makes the code simpler,
and, given that the compiler's code itself uses `mapIt`, also reduces
the amount of macro usage during bootstrapping.

Details
=======

* introduce the internal `mapIter` iterator
* remove the now obsolete internal `evalOnceAs` macro
* adjust the internal documentation of `mapIt` a bit
* adjust the `t8339.nim` test to not depend on `sequtils`
* don't introduce an unnecessary shallow copy when lifting `op` into a
  procedure

While the `mapIter` doesn't use `lent`, special care is taken for it to
be eligible for the cursor optimization. With the current compiler, for-
vars are always turned into cursors when the invoked inline iterator
only ever yields the *symbol* of a local.

The `t8339.nim` test is intended to make sure that aliases are skipped
when rendering types, but using `sequtils` for testing that is not
required, and only made the test unnecessarily depend on the `mapIt`
implementation details (which now changed, making the test fail).

zerbina committed:
sequtils: improve added documentation

As suggested by @saem.

saem commented:
/merge
